cpp demos logging unification

ivikhrev opened the pull request:
Unified demos logging. The following stages are left:

1. IE version
2. Model loading completion (with name and device)
3. IE parameters (num of streams, threads, requests)
4. Perfomance metrics.

Also done some improvements for text visualization (replaced putText with putHighlitedText in several places).

ivikhrev commented:
Jenkins please retry a build

ivikhrev commented:
Jenkins please retry a build

ivikhrev commented:
Jenkins please retry a build

vladimir-dudnik reviewed (changes requested):
as we agreed before, change report for version to:

[ INFO ] IE version: 2021.4.0
[ INFO ]    build: 2021.4.0-3839-cd81789d294-releases/2021/4

or, even better

[ INFO ] OpenVINO Inference Engine
[ INFO ]     version: 2021.4.0
[ INFO ]     build: 2021.4.0-3839-cd81789d294-releases/2021/4

add number of threads: auto where needed

G-API gaze_estimation_demo does not report num of ireqs/streams/threads at all
G-API interactive_face_detection_demo reports three times string [ INFO ] Network  is loaded  to CPU device without specifying model (seems when no model applied, info still printed, instead of model DISABLED string)

multi_channel_face_detection_demo does not print num of ireqs even though they were specified in cmd line

object_detection_demo prints string [ INFO ] Checking that the inputs are as the demo expects, which seems useless

G-API smart_classroom_demo does not report loaded models, nireqs, streams, threads at all

text_detection_demo reports string [ INFO ] Network C:\temp\models\intel\text-recognition-0012\FP16\text-recognition-0012.xml is loaded to CPU device. twice

ivikhrev commented:
> as we agreed before, change report for version to:
> 
> [ INFO ] IE version: 2021.4.0
> [ INFO ] build: 2021.4.0-3839-cd81789d294-releases/2021/4
> 
> or, even better
> 
> [ INFO ] OpenVINO Inference Engine
> [ INFO ] version: 2021.4.0
> [ INFO ] build: 2021.4.0-3839-cd81789d294-releases/2021/4
> 
Done.

> add number of threads: auto where needed
>
Done.

> G-API gaze_estimation_demo does not report num of ireqs/streams/threads at all
>
We can't get num of ireqs/streams/threads from G-API.
> G-API interactive_face_detection_demo reports three times string [ INFO ] Network is loaded to CPU device without specifying model (seems when no model applied, info still printed, instead of model DISABLED string)
> 
Done.
> multi_channel_face_detection_demo does not print num of ireqs even though they were specified in cmd line
> 
Fixed.
> G-API smart_classroom_demo does not report loaded models, nireqs, streams, threads at all
> 
Added info about loaded models. 
> text_detection_demo reports string [ INFO ] Network C:\temp\models\intel\text-recognition-0012\FP16\text-recognition-0012.xml is loaded to CPU device. twice
>
That's because model loading in this demo also happen twice. I didn't find the way to suppress second output.

eizamaliev commented on demos/classification_demo/cpp/main.cpp:
```diff
@@ -328,13 +325,23 @@ int main(int argc, char *argv[]) {
-        slog::info << slog::endl << "Metric reports:" << slog::endl;
+        slog::info << "Metric reports:" << slog::endl;
         metrics.printTotal();
+        slog::info << "Avg time:" << slog::endl;
```
```suggestion
```
It should be obvious. Also, `PerformanceMetrics` implementation could be changed.

ivikhrev replied:
Removed everywhere.

eizamaliev commented on demos/common/cpp/utils/include/utils/common.hpp:
```diff
@@ -365,3 +220,23 @@ inline std::string fileNameNoExt(const std::string &filepath) {
+
+inline void printExecNetworkInfo(const InferenceEngine::ExecutableNetwork& execNetwork, const std::string& modelName,
+    const std::string& deviceName) {
+    slog::info << "Network " << modelName << " is loaded to " << deviceName << " device." << slog::endl;
```
We should use common terms in all demos. 
@Wovchena @vladimir-dudnik @fzhar what should we use: `network`, `model`, `net` or other?
P.S.: most py/demos use the `model` term.

vladimir-dudnik replied:
`model` is ok for me

ivikhrev replied:
Changed.

eizamaliev commented on demos/gaze_estimation_demo/cpp_gapi/main.cpp:
```diff
@@ -164,6 +163,8 @@ int main(int argc, char *argv[]) {
             fileNameNoExt(FLAGS_m_fd) + ".bin",  // path to weights
             FLAGS_d_fd,                          // device specifier
         };
+        slog::info << "Network " << FLAGS_m_fd << " is loaded to " << FLAGS_d_fd << " device." << slog::endl;
```
When demo uses more than one model, it could be useful to print model type here, e.g.:
```
[ INFO ] The face detection model|network|net face.xml is loaded to CPU
```
and some errors would be more obvious:
```
[ INFO ] The face detection model|network|net age-gender-1337.xml is loaded to CPU
```

ivikhrev replied:
Added.

ivikhrev commented on demos/classification_demo/cpp/main.cpp:
```diff
@@ -328,13 +325,23 @@ int main(int argc, char *argv[]) {
 
         //// --------------------------- Report metrics -------------------------------------------------------
-        slog::info << slog::endl << "Metric reports:" << slog::endl;
+        slog::info << "Metric reports:" << slog::endl;
```
Done in commit [6d26b52](https://github.com/openvinotoolkit/open_model_zoo/pull/2547/commits/6d26b529d62b2245bf2da5eb5a5e0440b202c8cf)

ivikhrev commented on demos/classification_demo/cpp/main.cpp:
```diff
@@ -130,30 +126,29 @@ cv::Mat centerSquareCrop(const cv::Mat& image) {
         std::sort(imageNames.begin(), imageNames.end());
         for (size_t i = 0; i < imageNames.size(); i++) {
             const std::string& name = imageNames[i];
+            auto readingStart= std::chrono::steady_clock::now();
```
Done in commit [6d26b52](https://github.com/openvinotoolkit/open_model_zoo/pull/2547/commits/6d26b529d62b2245bf2da5eb5a5e0440b202c8cf)

ivikhrev commented on demos/common/cpp/utils/src/performance_metrics.cpp:
```diff
@@ -95,7 +95,6 @@ PerformanceMetrics::Metrics PerformanceMetrics::getTotal() const {
-    out << "Latency: " << std::fixed << std::setprecision(1) << metrics.latency << " ms\nFPS: " << metrics.fps << '\n';
-    std::cout << out.str();
+    slog::info << "\tLatency: " << std::fixed << std::setprecision(1) << metrics.latency << " ms" << slog::endl;
+    slog::info << "\tFPS : " << metrics.fps << slog::endl;
```
Done in commit [6d26b52](https://github.com/openvinotoolkit/open_model_zoo/pull/2547/commits/6d26b529d62b2245bf2da5eb5a5e0440b202c8cf)

anzhella-pankratova commented on demos/human_pose_estimation_demo/cpp/main.cpp:
```diff
@@ -343,8 +342,18 @@ int main(int argc, char *argv[]) {
-        slog::info << slog::endl << "Metric reports:" << slog::endl;
+        slog::info << "Metrics report:" << slog::endl;
         metrics.printTotal();
+        slog::info << "\tDecoding:\t" << std::fixed << std::setprecision(2) <<
```
kek
```suggestion
        slog::info << "\tDecoding:\t" << std::fixed << std::setprecision(2) <<
```

anzhella-pankratova replied:
The logging of preprocess/infer/postprocess metrics might be moved in a separate function, for example 
`void reportMetrics(ImagesCapture cap, AsyncPipeline pipeline, PerformanceMetrics renderMetrics)`

ivikhrev commented on demos/common/cpp/utils/include/utils/slog.hpp:
```diff
@@ -79,10 +81,20 @@ class LogStream {
+
+    // Specializing for std::vector and std::list
+    template<template<class, class> class Container, class T>
+    LogStream& operator<< (const Container<T, std::allocator<T>>& container) {
```
Implementation of `presenter.reportMeans()` was changed to allow print `[ INFO ]` on each new string in python demos also.  And in case of string we can't insert `slog::endl` into it for cpp demos.

ivikhrev commented on demos/common/cpp/utils/include/utils/slog.hpp:
```diff
@@ -79,10 +81,20 @@ class LogStream {
 
 
 static LogStream info("INFO", std::cout);
+static LogStream dbg("DEBUG", std::cout);
```
Changed.

ivikhrev commented on demos/common/cpp/utils/include/utils/common.hpp:
```diff
@@ -365,3 +220,28 @@ inline std::string fileNameNoExt(const std::string &filepath) {
+            }
+        }
+    }
+    catch (...) {}
```
Specified ` InferenceEngine::Exception`.

ivikhrev commented on demos/common/cpp/utils/include/utils/slog.hpp:
```diff
@@ -66,6 +67,7 @@ class LogStream {
         return *this;
     }
 
+
```
Ok.

Wovchena commented on demos/common/python/monitors_extension/monitors_extension.cpp:
```diff
@@ -96,7 +96,12 @@ PyObject *presenter_reportMeans(PresenterObject *self, PyObject *Py_UNUSED(ignor
+        for(const auto& s : report){
+            PyList_Append(PList, Py_BuildValue("s", s.c_str()));
+        }
+        return PList;
```
I believe this should be in @anzhella-pankratova's patch because that's where people pay attention to how python prints are formatted.

Wovchena replied:
But you need `reportMeans()` to return a vector... Never mind

Wovchena commented on demos/common/cpp/utils/include/utils/common.hpp:
```diff
@@ -365,3 +220,31 @@ inline std::string fileNameNoExt(const std::string &filepath) {
     return filepath.substr(0, pos);
 }
+
+inline void printExecNetworkInfo(const InferenceEngine::ExecutableNetwork& execNetwork, const std::string& modelName,
```
You should align on on implementations of `printExecNetworkInfo()` and `log_runtime_settings()` with @anzhella-pankratova.

ivikhrev replied:
we aligned the way nstreams, threads and infer requests are reported [(python impl)](https://github.com/openvinotoolkit/open_model_zoo/pull/2567/commits/a775b5d3d7c59cca9a70e41af430a629b522915e). Is it ok right now?

Wovchena replied:
Yes

anzhella-pankratova commented on demos/common/cpp/pipelines/src/async_pipeline.cpp:
```diff
@@ -36,8 +34,8 @@ AsyncPipeline::AsyncPipeline(std::unique_ptr<ModelBase>&& modelInstance, const C
                 "OPTIMAL_NUMBER_OF_INFER_REQUESTS ExecutableNetwork metric. Failed to query the metric with error: ") + ex.what());
         }
     }
+    slog::info << "\tNumber of network inference requests: " << nireq << "." << slog::endl;
```
```suggestion
    slog::info << "\tNumber of network inference requests: " << nireq << slog::endl;
```

ivikhrev replied:
Thanks!

ivikhrev commented on demos/common/cpp/utils/include/utils/common.hpp:
```diff
@@ -365,3 +220,27 @@ inline std::string fileNameNoExt(const std::string &filepath) {
+                slog::info << "\t\tNumber of streams: " << nstreams << slog::endl;
+                if (device == "CPU") {
+                    std::string nthreads = execNetwork.GetConfig("CPU_THREADS_NUM").as<std::string>();
+                    slog::info << "\t\tNumber of threads: " << (nthreads == "0" ? "AUTO" : nthreads) <<slog::endl;
```
```suggestion
                    slog::info << "\t\tNumber of threads: " << (nthreads == "0" ? "AUTO" : nthreads) << slog::endl;
```

ivikhrev commented on demos/common/cpp/utils/include/utils/common.hpp:
```diff
@@ -365,3 +220,27 @@ inline std::string fileNameNoExt(const std::string &filepath) {
+
+inline void printExecNetworkInfo(const InferenceEngine::ExecutableNetwork& execNetwork, const std::string& modelName,
+    const std::string& deviceName, const std::string& modelType = "") {
+    slog::info << "The " << modelType << (modelType.empty() ? "" : " ")  << "model " << modelName << " is loaded to " << deviceName << slog::endl;
```
```suggestion
    slog::info << "The " << modelType << (modelType.empty() ? "" : " ") << "model " << modelName << " is loaded to " << deviceName << slog::endl;
```

ivikhrev commented:
Jenkins please retry a build

ivikhrev commented:
Jenkins please retry a build

fzhar commented on demos/common/cpp/utils/include/utils/config_factory.h:
```diff
@@ -16,11 +16,13 @@
 struct CnnConfig {
-    std::string devices;
+    std::string deviceName;
+    std::set<std::string> devices;
```
we'd better put devices to private or protected section with public getter - to avoid setting devices explicetely

ivikhrev replied:
Ok, made it as protected.

vladimir-dudnik commented:
@ivikhrev please rebase

ivikhrev commented:
@vladimir-dudnik it's ok now
